expr/seriesaggregators.go

```diff
@@ -126,3 +137,125 @@ func crossSeriesSum(in []models.Series, out *[]schema.Point) {
 		*out = append(*out, point)
 	}
 }
+
+func crossSeriesMultiply(in []models.Series, out *[]schema.Point) {
+	for i := 0; i < len(in[0].Datapoints); i++ {
+		product := float64(1)
+		for j := 0; j < len(in); j++ {
+			p := in[j].Datapoints[i].Val
+			if math.IsNaN(p) {
+				// NaN * anything equals NaN()
+				product = math.NaN()
+				break
+			}
+			product *= p
+		}
+		point := schema.Point{
+			Ts:  in[0].Datapoints[i].Ts,
+			Val: product,
+		}
+
+		*out = append(*out, point)
+	}
+}
+
+func crossSeriesMedian(in []models.Series, out *[]schema.Point) {
+	vals := make([]float64, 0, len(in))
+	for i := 0; i < len(in[0].Datapoints); i++ {
+		vals = vals[:0]
+		for j := 0; j < len(in); j++ {
+			p := in[j].Datapoints[i].Val
+			if !math.IsNaN(p) {
+				vals = append(vals, p)
+			}
+		}
+		point := schema.Point{
+			Ts: in[0].Datapoints[i].Ts,
+		}
+		if len(vals) == 0 {
+			point.Val = math.NaN()
+		} else {
+			sort.Float64s(vals)
+			mid := len(vals) / 2
+			if len(vals)%2 == 0 {
+				point.Val = (vals[mid-1] + vals[mid]) / 2
+			} else {
+				point.Val = vals[mid]
+			}
+		}
+
+		*out = append(*out, point)
+	}
+}
+
+func crossSeriesDiff(in []models.Series, out *[]schema.Point) {
+	for i := 0; i < len(in[0].Datapoints); i++ {
+		val := math.NaN()
+		for j := 0; j < len(in); j++ {
+			p := in[j].Datapoints[i].Val
+			if !math.IsNaN(p) {
+				if math.IsNaN(val) {
+					val = p
+				} else {
+					val -= p
+				}
+			}
+		}
+		point := schema.Point{
+			Ts:  in[0].Datapoints[i].Ts,
+			Val: val,
+		}
+
+		*out = append(*out, point)
+	}
+}
+
+func crossSeriesStddev(in []models.Series, out *[]schema.Point) {
+	averages := make([]schema.Point, 0, len(in[0].Datapoints))
+	crossSeriesAvg(in, &averages)
+
+	for i := 0; i < len(in[0].Datapoints); i++ {
+		point := schema.Point{
+			Ts:  in[0].Datapoints[i].Ts,
+			Val: math.NaN(),
+		}
+
+		if !math.IsNaN(averages[i].Val) {
+			num := float64(0)
+			totalDeviationSquared := float64(0)
+			for j := 0; j < len(in); j++ {
+				p := in[j].Datapoints[i].Val
+				if !math.IsNaN(p) {
+					num++
+					deviation := p - averages[i].Val
+					totalDeviationSquared += deviation * deviation
+				}
+			}
+
+			point.Val = math.Sqrt(totalDeviationSquared / num)
+		}
+
+		*out = append(*out, point)
+	}
+}
+
+func crossSeriesRange(in []models.Series, out *[]schema.Point) {
+	maxes := make([]schema.Point, 0, len(in[0].Datapoints))
+	mins := make([]schema.Point, 0, len(in[0].Datapoints))
+
+	crossSeriesMax(in, &maxes)
+	crossSeriesMin(in, &mins)
+
+	for i := 0; i < len(in[0].Datapoints); i++ {
+		point := schema.Point{
+			Ts:  in[0].Datapoints[i].Ts,
+			Val: math.NaN(),
+		}
+
+		if !math.IsNaN(maxes[i].Val) {
```

Comment:
is this check needed? seems like the subtraction would do the right thing in all cases. https://play.golang.org/p/hiROasfqyq

Reply:
Apparently not! Cool.